### PR TITLE
[feature/prevent-sleep-detail-view] Presentation Mode

### PR DESCRIPTION
**Pull request by hosy**
## Description
Added an action in detail view menu which enables presentation mode.
Presentation mode prevents the display from sleep mode as long as the detail view is dismissed. Furthermore the navigation bar will be hidden and the file opens in full screen.

## Related Issue
#704 

## Motivation and Context
This could be very useful when a user wants to present a file or on a talk.

## How Has This Been Tested?
- open item in detail view
- open the more menu in the navigation bar
- select `Presentation Mode` and `Enable`
- be sure the device will not go into sleep mode (needs to be tested on a real device)

## Screenshots (if appropriate):
| Simulator Screen Shot - iPhone 11 Pro Max - 2021-02-12 at 17 43 36 | Simulator Screen Shot - iPhone 11 Pro Max - 2021-02-12 at 17 43 47 |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Added changelog files for the fixed issues in folder changelog/unreleased

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%2011.6/Presentation%20mode.md

Bugs & improvements:

- [x] (1) Presentation mode in non-supported files https://github.com/owncloud/ios-app/pull/893#issuecomment-791405724
- [x] (2) txt & MSO can't escape from fill screen mode https://github.com/owncloud/ios-app/pull/893#issuecomment-791413512

### CODE REVIEW

File: changelog/unreleased/704 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Change: Presentation Mode
+
+Added an action in detail view menu which enables presentation mode.
+Presentation mode prevents the display from sleep mode as long as the detail view is dismissed. Furthermore the navigation bar will be hidden.
```
Comment (michaelstingl) [inline changelog/unreleased/704:4]:
```suggestion
Presentation mode prevents the display from sleep mode as long as the detail view is closed. Furthermore the navigation bar will be hidden.
```

Not sure if "dismissed" is the common term?

File: ownCloud/Client/Actions/Actions+Extensions/PresentationModeAction.swift [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	// MARK: - Action implementation
+	override func run() {
+		guard context.items.count > 0, let hostViewController = context.viewController else {
+			completed(with: NSError(ocError: .insufficientParameters))
+			return
+		}
+
+		let alertController = UIAlertController(title: "Presentation Mode".localized, message: "Enabling presentation mode will prevent the display from sleep mode until the view is dismissed.".localized, preferredStyle: .alert)
```
Comment (michaelstingl) [inline ownCloud/Client/Actions/Actions+Extensions/PresentationModeAction.swift:47]:
```suggestion
		let alertController = UIAlertController(title: "Presentation Mode".localized, message: "Enabling presentation mode will prevent the display from sleep mode until the view is closed.".localized, preferredStyle: .alert)
```

See comment in `changelog/unreleased/704`

**Review by michaelstingl — COMMENTED** [review]
unsure about the usage of "dismissed" vs. "closed"

**Comment by hosy** [comment]
`Close` also works for me!

File: ownCloud/Client/Actions/Actions+Extensions/PresentationModeAction.swift (hunk 2 above)
Comment (hosy) [reply ownCloud/Client/Actions/Actions+Extensions/PresentationModeAction.swift]: Fixed wording!

**Review by felix-schwarz — CHANGES_REQUESTED** [review]
Code-wise I noticed that `DisplaySleepPreventer`'s existing method set was not covering a use case like Presentation Mode too well. F.ex. if the OC app is opened in two windows on an iPad, side-by-side, and one half shows the Status tab (which also prevents sleep), then the Presentation Mode action would not prompt the user - but leaving the viewer would end sleep prevention; even if Presentation Mode wasn't used.

Since I wrote `DisplaySleepPreventer` I took the liberty to extend it a bit, so that it supports "reasons" for preventing sleep, allowing Presentation Mode to track its status independently.

During testing, I found another issue: if I present `Example.odt` (or any other file for which there is no viewer support) and then select Presentation Mode, I end up with no way to return:
Simulator Screen Shot - iPhone 12 mini - 2021-02-24 at 00 33 15

**Comment by hosy** [comment]
@felix-schwarz just pushed a new commit to prevent fullscreen support on unsupported subclasses. Please give it a new try. Thanks!

**Comment by felix-schwarz** [comment]
@hosy Thanks. That solves the issue nicely.

**Review by felix-schwarz — APPROVED** [review]

**Comment by jesmrec** [comment]
### (1)

About your last finding (open Presentation mode for non-supported files)

Would it make sense to remove the option from there? in case there is no file to display, presentation mode is not needed

**Comment by jesmrec** [comment]
### (2)

1. Open a txt file or a MS Office file (doc, xls, ppt...)
2. Open Menu and select `Presentation Mode`

Current: Preview turns to full screen and there is no way to dismiss it. Other kind of files like PDF or pictures work fine (full screen is dismissed with a tap or a double tap).

Expected: txt and MSOffice files can prevent the full screen mode

iPhoneXR v14.4
`e1b6fb9`

**Comment by hosy** [comment]
@jesmrec regarding (1)
You mean to do only show the menu item, if the file is locally available? 
Could make sense!

**Comment by hosy** [comment]
@jesmrec regarding (2)
It would make sense to show this file types in fullscreen too. I will check if I can add a tap handler to exit the fullscreen mode.

**Comment by jesmrec** [comment]
> @jesmrec regarding (1)
> You mean to do only show the menu item, if the file is locally available?
> Could make sense!

yes, and also, if the file is not previewable (f. ex., a zip file), it will be never presented so it is not needed to be shown.

**Comment by jesmrec** [comment]
(1) and (2) fixed

2nd round finished.

Approved on my side

**Review by jesmrec — APPROVED** [review]

**Pull request merged**